/* Copyright 2023 The FedLearner Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

//ref: https://github.com/bytedance/fedlearner/blob/master/protocols/fedlearner/common/tree_model.proto

syntax = "proto3";

package fedlearner.common;

message RegressionTreeNodeProto {
    int32 node_id = 1;
    int32 left_child = 2;
    int32 right_child = 3;
    int32 parent = 4;
    bool is_owner = 5;
    int32 owner_id = 6;
    int32 feature_id = 7;
    float threshold = 8;
    float weight = 9;
    bool default_left = 10;
    bool is_cat_feature = 11;
    repeated int32 cat_threshold = 12;
}

message RegressionTreeProto {
    repeated RegressionTreeNodeProto nodes = 1;
    repeated float feature_importance = 3;
};

message BoostingTreeEnsambleProto {
    repeated RegressionTreeProto trees = 1;
    repeated float feature_importance = 2;
    repeated string feature_names = 3;
    repeated string cat_feature_names = 4;
};
